src/core/search/indices.cc

```diff
@@ -61,18 +67,43 @@ vector<DocId> BaseStringIndex::Matching(string_view str) const {
 }
 
 void TextIndex::Add(DocId doc, string_view value) {
-  for (const auto& word : GetWords(value)) {
+  for (const auto& word : Tokenize(value)) {
     auto& list = entries_[word];
     list.insert(upper_bound(list.begin(), list.end(), doc), doc);
   }
 }
 
+void TextIndex::Remove(DocId doc, string_view value) {
+  for (const auto& word : Tokenize(value)) {
+    auto& list = entries_[word];
+    auto it = lower_bound(list.begin(), list.end(), doc);
+    if (it != list.end() && *it == doc)
+      list.erase(it);
+  }
+}
+
 void TagIndex::Add(DocId doc, string_view value) {
   auto tags = absl::StrSplit(value, ',');
   for (string_view tag : tags) {
-    auto& list = entries_[absl::StripAsciiWhitespace(tag)];
+    tmp_buf_ = absl::StripAsciiWhitespace(tag);
+    absl::AsciiStrToLower(&tmp_buf_);
+
+    auto& list = entries_[tmp_buf_];
     list.insert(upper_bound(list.begin(), list.end(), doc), doc);
   }
 }
 
+void TagIndex::Remove(DocId doc, string_view value) {
+  auto tags = absl::StrSplit(value, ',');
+  for (string_view tag : tags) {
+    tmp_buf_ = absl::StripAsciiWhitespace(tag);
+    absl::AsciiStrToLower(&tmp_buf_);
```

Comment:
Now this transform part is duplicated... But implementing some kind of transforming iterator is unfortunately even more cumbersome

Reply:
Maybe just add a `Normalize()` method?

Reply:
Added. Even though it adds allocations, we can filter duplicate tags there for safety